### PR TITLE
Functional tests for WP-CLI commands

### PR DESCRIPTION
**Pull request by swissspidy**
### Description of the Change

This PR sets up the necessary infrastructure for running Behat tests for the built-in WP-CLI commands.

Also sets up code coverage, which is quite convenient.

### How to test the Change

Run the tests locally or check out the CI build status.

### Changelog Entry

> Added - Functional tests for WP-CLI command

### Checklist:

- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my change.
- [x] All new and existing tests pass.

### CODE REVIEW

**Review by mukeshpanchal27 — COMMENTED** [review]
Thank you, @swissspidy, for the PR. I have reviewed it initially and would like to discuss a few questions:

- Do we have any plans to add new tests in Behat?
- Considering that Behat uses Gherkin syntax, there might be a learning curve for us. How do we plan to address this?
- Are we going to maintain both Unit tests and Behat tests simultaneously?

**Comment by swissspidy** [comment]
Great questions!

> * Do we have any plans to add new tests in Behat?

Right now this PR just lays the foundation for the tests. New tickets can be created to add a few more tests. Not too many tests are needed because the command is rather straightforward and only has a few options.

> * Considering that Behat uses Gherkin syntax, there might be a learning curve for us. How do we plan to address this?

Luckily the syntax is pretty easy to grasp, making it easy to understand and write tests, plus there's plenty of documentation and tutorials out there. And of course I'm happy to do an intro as well if there's interest.

> * Are we going to maintain both Unit tests and Behat tests simultaneously?

Yes, these tests are an addition to the existing PHPUnit-based tests in the repo, focusing on how users interact with the WP-CLI command. There is little overlap. Just like e2e tests and unit/integration tests are not mutually exclusive. So I guess these could be seen as e2e tests for the WP-CLI command :-)

File: tests/behat/features/plugin-check.feature [hunk 1]
```diff
@@ -0,0 +1,10 @@
+Feature: Test that the WP-CLI command works.
+
+  Scenario: Check a non-existent plugin
```
Comment (spacedmonkey) [inline tests/behat/features/plugin-check.feature:3]: I really wish we had a test of a plugin that works. Just testing failure here feels weird.

Comment (swissspidy) [reply tests/behat/features/plugin-check.feature:3]: Right now this PR just lays the foundation for the tests. New tickets can be created to add a few more tests.

Comment (spacedmonkey) [reply tests/behat/features/plugin-check.feature:3]: How can we ensure the foundation work if the tests do not do anything meaningful?

Comment (swissspidy) [reply tests/behat/features/plugin-check.feature:3]: There was lots of manual testing of the CLI command so far, and the tests setup is standard for WP-CLI and battle-tested. Are you looking for more?

**Comment by swissspidy** [comment]
@felixarntz @westonruter @joemcgill @mukeshpanchal27 Pinging you since this PR has been around for a long time but it's still valuable — this adds some initial Behat tests for the WP-CLI command added by this plugin.

We need this to ensure the command works properly, especially when making changes. Also with #153 in mind.

File: .github/workflows/behat-test.yml [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    strategy:
+      fail-fast: true
+      matrix:
+        php:
+        - '7.0'
+        - '7.1'
+        - '7.2'
+        - '7.3'
+        - '7.4'
+        - '8.1'
```
Comment (mukeshpanchal27) [inline .github/workflows/behat-test.yml:68]:
```suggestion
        - '8.0'
        - '8.1'
```

File: .github/workflows/behat-test.yml [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        - '8.2'
+        wordpress: [ 'latest' ]
+        include:
+          - php: '8.0'
+            wordpress: 'latest'
+            coverage: true
+          - php: '8.0'
+            wordpress: '6.0'
+          - php: '8.0'
+            wordpress: '6.1'
```
Comment (mukeshpanchal27) [inline .github/workflows/behat-test.yml]:
```suggestion
```

In #292, we set the minimum WordPress version for the plugin checker to be 6.3. Let's remove this old WP version check.

File: .github/workflows/behat-test.yml (hunk 2 above)
Comment (swissspidy) [reply .github/workflows/behat-test.yml:68]: 8.0 with WP latest is already added below, in the `include` section, with code coverage enabled.

File: .github/workflows/behat-test.yml (hunk 3 above)
Comment (swissspidy) [reply .github/workflows/behat-test.yml]: Ah right, I'll set this to 6.3 then.

**Review by mukeshpanchal27 — APPROVED** [review]
Thanks @swissspidy, LGTM.

File: tests/behat/includes/FeatureContext.php [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+	/**
+	 * @When /^I (run|try) the WP-CLI command `([^`]+)`$/
+	 */
+	public function when_i_run_the_wp_cli_command( $mode, $command ) {
+		$command = "wp {$command}";
+
+		$with_code_coverage = getenv( 'BEHAT_CODE_COVERAGE' );
+		if ( \in_array( $with_code_coverage, [ true, 'true', 1, '1' ], true ) ) {
+			$command = "{$command} --require={PROJECT_DIR}/tests/behat/utils/maybe-generate-wp-cli-coverage.php";
```
Comment (felixarntz) [inline tests/behat/includes/FeatureContext.php:93]: Does this work when there is an existing `--require` argument? That would be relevant for runtime checks due to the current limitations of having to require the early logic.

**Review by felixarntz — COMMENTED** [review]
@swissspidy LGTM, just one question before merge.

File: tests/behat/includes/FeatureContext.php (hunk 4 above)
Comment (swissspidy) [reply tests/behat/includes/FeatureContext.php:93]: Yes, WP-CLI supports multiple `--require` arguments.

Also worth noting that this kind of `--require` usage should become obsolete anyway with #153

Comment (swissspidy) [reply tests/behat/includes/FeatureContext.php:93]: @felixarntz Please let me know if this answers your question :-)

Comment (felixarntz) [reply tests/behat/includes/FeatureContext.php:93]: Thanks, LGTM!

**Review by felixarntz — APPROVED** [review]

**Pull request merged**